### PR TITLE
Improve UI

### LINKED ISSUE #117: Multiple configuration files are not supported
detekt supports multiple configuration files, but the plugin only supports one. I have split my detekt configuration into multiple files: one shared across all projects and one project-specific. So, I cannot use the plugin.

### LINKED ISSUE #135: Store configuration file path as relative to the project root
Right now, as of v1.17.1, the Detekt configuration path is stored as an absolute path. Given that the Detekt IJ plugin settings are per-project, and thus saved in `.idea/detekt.xml`, this prevents folks from sharing the Detekt plugin configuration with other devs by committing the `.idea/detekt.xml` file as that contains absolute paths which will only work for one user.

This would easily be solved by using relative paths instead, referred to the project root directory.

### LINKED ISSUE #231: File paths should be platform-independent
Config file paths (and possibly other paths such as baseline) are stored in `detekt.xml` and read back as platform-dependent.

This means that if the `detekt.xml` is [shared in VCS](https://blog.sebastiano.dev/share-the-settings-with-the-whole-team-a-year-later/) for the whole team to use, and not everyone is on the same OS (mix of POSIX and Windows users) the paths will not work for some folks and they'll be bombarded by "file not found" notifications from the Detekt plugin.

The plugin should store paths in a platform-independent way and apply the correct separators at runtime.

### PR DESCRIPTION
**Pull request by rock3r**
This PR implements an all-new UI for the plugin configuration and solves a number of issues.

The new UI is created using the Kotlin UI DSL; since our minimum supported version is 2020.3, I needed to implement the UI _twice_, both in the v1 and v2 DSL, for IJ < 21.3 and >= 21.3 respectively. The UIs are almost indistinguishable from each other from a user's perspective (but this required some hacking around on both DSLs).

I have added explanations (with links to detekt.dev) for the file-related settings, and made it possible to have multiple config files and plugins, with a decent UI. The paths are also treated system-independently and thus they are correctly relativized by the IDE in the settings xml file.

The settings file has changed in format, and there is a best-effort attempt at migrating from the old to the new settings format.

 v1 DSL (legacy version for IJ < 2021.3) | v2 DSL (for IJ >= 2021.3)
 --- | ---
 v1  | v2 screenshot

> **Note**: as mentioned, there's a number of hacks involved; the V1 ones are "forever" at this point, I hope the V2 ones will get better over time as we raise the minimum IDE version (21.3 is the next version that would make a difference, dropping the v1 DSL UI entirely; then 21.1 for some API changes)

> **Note**: I have developed this under Windows. I haven't run this under Linux or macOS, although I don't expect this to cause issues.

>**Warning**: I can't get the tests to run on my machine in any way, so I didn't touch those. They will likely need a pass to get them to work (although they do compile at least)

>**Warning**: I have tested this new UI and it works fine, but I think it would be great if y'all folks gave it a test to see if I broke anything anywhere

Closes #231
Closes #135
Closes #117

### CODE REVIEW

**Comment by rock3r** [comment]
@arturbosch can you approve the workflow to run?

**Comment by arturbosch** [comment]
> @arturbosch can you approve the workflow to run?

Wow, screenshots look really good.
I approved the CI. Will review the PR later today 👍🏼

**Comment by rock3r** [comment]
I have some minor tweaks to do to the v2 side of things after talking to folks on how to improve a couple of things, but should be pretty minor. I can get rid of a hack

**Comment by rock3r** [comment]
I'll review what's wrong with the CI shortly

**Comment by rock3r** [comment]
@arturbosch as expected tests are red, but unfortunately I can't see what failed because the test results output is not archived. Have you considered something like [this](https://github.com/marketplace/actions/junit-report-action)?

File: build.gradle.kts [hunk 1]
```diff
@@ -11,7 +11,7 @@ repositories {
 }
 
 plugins {
-    id("org.jetbrains.intellij").version("1.6.0")
```
Comment (arturbosch) [inline build.gradle.kts:14]: The test task is broken with 1.8.0 and/or Gradle 7.5.x.
Java 11, Gradle 7.4.2 and plugin 1.6.0 works.
I think you also run into https://github.com/detekt/detekt-intellij-plugin/issues/211

File: src/main/kotlin/io/gitlab/arturbosch/detekt/idea/util/PluginUtils.kt [hunk 2]
```diff
@@ -16,4 +19,13 @@ object PluginUtils {
         }
         return plugin.version
     }
+
+    fun isAtLeastIJBuild(minBuild: String): Boolean =
+        ApplicationInfo.getInstance().build > BuildNumber.fromString(minBuild)!!
+
+    fun Set<String>.toVirtualFilesList(): List<VirtualFile> {
```
Comment (arturbosch) [inline src/main/kotlin/io/gitlab/arturbosch/detekt/idea/util/PluginUtils.kt:26]: please move this to `ProjectUtils`

**Comment by arturbosch** [comment]
> @arturbosch as expected tests are red, but unfortunately I can't see what failed because the test results output is not archived. Have you considered something like [this](https://github.com/marketplace/actions/junit-report-action)?

I can confirm. Reverting the intellij plugin to 1.6.0 works.

**Comment by arturbosch** [comment]
Please take a look at the attached video. It seems that resizing is a bit off for the page now for the lists

[Screencast from 08-08-2022 04:00:58 PM.webm](https://user-images.githubusercontent.com/20924106/183436315-6fb8c02f-3968-4464-96fe-54271dbbe36c.webm)

File: build.gradle.kts (hunk 1 above)
Comment (rock3r) [reply build.gradle.kts:14]: Ahhhhhhh damn. Let me roll back Gradle to 7.4.2 and see if that's enough :)

File: src/main/kotlin/io/gitlab/arturbosch/detekt/idea/config/ui/LegacyDetektConfigUiProvider.kt [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+import javax.swing.JComponent
+import javax.swing.JLabel
+
+@Suppress("DialogTitleCapitalization") // It gets tripped up by the capitalization of Detekt's name
+class LegacyDetektConfigUiProvider(
+    private val settings: DetektPluginSettings,
+    private val project: Project
+) : DetektConfigUiProvider {
+
+    @Suppress("DEPRECATION") // This legacy UI is only used on IJ < 21.3 where the v2 DSL is unavailable
```
[the remark this answers is not in the document]
Comment (rock3r) [reply src/main/kotlin/io/gitlab/arturbosch/detekt/idea/config/ui/LegacyDetektConfigUiProvider.kt:25]: The issue is we need to keep maintaining the old DSL for IJ 21.2 and lower, it makes no difference unfortunately to raise the minimum version to 21.2 as we'd need to raise to 21.3 to get rid of it. Once Dolphin goes stable we can drop the V1 UI and raise the minimum supported version to 2021.3... Shouldn't be too bad as I don't expect we'll need to make too many changes in the medium term, but we can't wait the X months it'll take for Dolphin to go stable.

**Comment by arturbosch** [comment]
> This PR implements an all-new UI for the plugin configuration and solves a number of issues.
> 
> The new UI is created using the Kotlin UI DSL; since our minimum supported version is 2020.3, I needed to implement the UI _twice_, both in the v1 and v2 DSL, for IJ < 21.3 and >= 21.3 respectively. The UIs are almost indistinguishable from each other from a user's perspective (but this required some hacking around on both DSLs).
> 
> I have added explanations (with links to detekt.dev) for the file-related settings, and made it possible to have multiple config files and plugins, with a decent UI. The paths are also treated system-independently and thus they are correctly relativized by the IDE in the settings xml file.
> 
> The settings file has changed in format, and there is a best-effort attempt at migrating from the old to the new settings format.
> 
> v1 DSL (legacy version for IJ < 2021.3)	v2 DSL (for IJ >= 2021.3)
> v1	v2 screenshot
> > Note: as mentioned, there's a number of hacks involved; the V1 ones are "forever" at this point, I hope the V2 ones will get better over time as we raise the minimum IDE version (21.3 is the next version that would make a difference, dropping the v1 DSL UI entirely; then 21.1 for some API changes)
> 
> > Note: I have developed this under Windows. I haven't run this under Linux or macOS, although I don't expect this to cause issues.
> 
> > Warning: I can't get the tests to run on my machine in any way, so I didn't touch those. They will likely need a pass to get them to work (although they do compile at least)
> 
> > Warning: I have tested this new UI and it works fine, but I think it would be great if y'all folks gave it a test to see if I broke anything anywhere


Works on Linux on IJ 2022.2 and AS Chipmunk. 👍🏼 
However the `apply` button does not work for the checkboxes, only when configuration file is added.
Therefore the setting changes are not applied.

File: build.gradle.kts (hunk 1 above)
Comment (rock3r) [reply build.gradle.kts:14]: Yep it's something with Gradle 7.5, rolled back to 7.4.2 and it works. Can keep using the IJ plugin 1.8.0.

File: src/main/kotlin/io/gitlab/arturbosch/detekt/idea/util/ProjectUtils.kt [hunk 4]
```diff
@@ -46,6 +48,8 @@ fun showNotification(title: String, content: String, project: Project) {
         content,
         NotificationType.WARNING
     )
+
+    @Suppress("DialogTitleCapitalization")
     notification.addAction(object : AnAction("Open Detekt projects settings") {
```
[the remark this answers is not in the document]
Comment (rock3r) [reply src/main/kotlin/io/gitlab/arturbosch/detekt/idea/util/ProjectUtils.kt:53]: Rephrased and took the opportunity to extract the strings in this file to the properties file.

**Comment by rock3r** [comment]
> I can confirm. Reverting the intellij plugin to 1.6.0 works.
This didn't work for me, but downgrading Gradle to 7.4.2 did the trick :) Will be pushing soon.

**Comment by rock3r** [comment]
As for the apply/reset button: yeah there's something that's missing there. Let me look into it. I'll push the rest of the fixes in the meantime.

**Comment by rock3r** [comment]
> Please take a look at the attached video. It seems that resizing is a bit off for the page now for the lists
> 
> [Screencast from 08-08-2022 04:00:58 PM.webm](https://user-images.githubusercontent.com/20924106/183436315-6fb8c02f-3968-4464-96fe-54271dbbe36c.webm) Screenshot from 2022-08-08 16-04-55

Yeah the minimum width is enforced by the UI DSL/list component itself, not something I have set. It looks like it's a bug with the DSL itself as this other page is also done with the DSL and exhibits the same behaviour:

image

The page you were comparing to is not using the DSL, so it doesn't have the same issue.

**Comment by arturbosch** [comment]
> > Please take a look at the attached video. It seems that resizing is a bit off for the page now for the lists
> > [Screencast from 08-08-2022 04:00:58 PM.webm](https://user-images.githubusercontent.com/20924106/183436315-6fb8c02f-3968-4464-96fe-54271dbbe36c.webm) Screenshot from 2022-08-08 16-04-55
> 
> Yeah the minimum width is enforced by the UI DSL/list component itself, not something I have set. It looks like it's a bug with the DSL itself as this other page is also done with the DSL and exhibits the same behaviour:
> 
> image
> 
> The page you were comparing to is not using the DSL, so it doesn't have the same issue.

Can we somehow move the title just above the list like in this screenshot (Manage standalone scripts)?
I think then the alignment should be fixed. Or we wait for IJ to fix it ^^

**Comment by rock3r** [comment]
I asked my colleague who works on the DSL if there's a fix for this — you can see the Swagger settings have the same issue without having a label on the side :)

**Comment by rock3r** [comment]
FYI I tested without a label on the side and it makes no difference, it's definitely a DSL bug.

**Comment by arturbosch** [comment]
> As for the apply/reset button: yeah there's something that's missing there. Let me look into it. I'll push the rest of the fixes in the meantime.

👍🏼 ping me to test the apply changes and it's ready to merge (release)

**Comment by rock3r** [comment]
I think I fixed the apply/reset issues. I've tested on 22.1, need to test on 20.3 too but it should be ok.

File: src/main/kotlin/io/gitlab/arturbosch/detekt/idea/config/DetektConfig.kt [hunk 5]
```diff
@@ -1,28 +1,35 @@
...
-
-    override fun isModified(): Boolean = !detektConfigurationForm.isNotModified
-
-    override fun getId(): String = "io.github.detekt.config"
-
-    override fun getDisplayName(): String = DETEKT
-
-    override fun apply() {
-        detektConfigurationForm.apply()
-        DaemonCodeAnalyzer.getInstance(project).restart()
```
Comment (arturbosch) [inline src/main/kotlin/io/gitlab/arturbosch/detekt/idea/config/DetektConfig.kt:22]: This line is important to reanalyze when changing any detekt settings.
You can open a file with an issue and change the `warningsAsErrors` setting and notice that the file is not analyzed.

**Comment by arturbosch** [comment]
> I think I fixed the apply/reset issues. I've tested on 22.1, need to test on 20.3 too but it should be ok.

works now 👍🏼 

Thanks for the awesome work! Please see my last comment and we are rdy to merge.

File: src/main/kotlin/io/gitlab/arturbosch/detekt/idea/config/DetektConfig.kt (hunk 5 above)
Comment (rock3r) [reply src/main/kotlin/io/gitlab/arturbosch/detekt/idea/config/DetektConfig.kt:22]: Ah I forgot to put that back in! Will do asap

**Comment by rock3r** [comment]
I did test on 20.3 as part of the last few bugfixes. I'm pretty happy with how this turned out :)

Will get the last thing done as soon as I have 5 min later today

**Comment by rock3r** [comment]
All done :)

**Review by arturbosch — APPROVED** [review]

**Comment by rock3r** [comment]
:partyparrot:

**Pull request merged**